feature-libs/cart/base/core/facade/active-cart.service.ts

```diff
@@ -378,54 +378,63 @@ export class ActiveCartService implements ActiveCartFacade, OnDestroy {
-      filter(([cartState, userId]) =>
-        Boolean(
-          userId === OCC_USER_ID_ANONYMOUS ||
-            cartState.success ||
-            cartState.error
-        )
-      ),
-      take(1),
-      tap(([cartState, userId]) => {
-        if (isEmpty(cartState.value)) {
-          this.multiCartFacade.createCart({
-            userId,
-            extraData: {
-              active: true,
-            },
-          });
-        }
-      }),
-      switchMapTo(cartSelector$),
-      filter((cartState) => cartState.success || cartState.error),
-      // wait for active cart id to point to code/guid to avoid some work on temp cart entity
-      withLatestFrom(this.activeCartId$),
-      filter(([cartState, cartId]) => !this.isCartCreating(cartState, cartId)),
-      map(([cartState]) => cartState.value),
-      filter((cart) => !isEmpty(cart)),
-      take(1)
-    );
+    return this.activeCartId$
+      .pipe(
+        // Avoid load/create call when there are new cart creating at the moment
+        withLatestFrom(cartSelector$),
+        filter(
+          ([cartId, cartState]) => !this.isCartCreating(cartState, cartId)
+        ),
+        map(([, cartState]) => cartState),
+        take(1)
+      )
+      .pipe(
+        withLatestFrom(this.userIdService.getUserId()),
+        tap(([cartState, userId]) => {
+          // Try to load the cart, because it might have been created on another device between our login and add entry call
+          if (
+            isEmpty(cartState.value) &&
+            userId !== OCC_USER_ID_ANONYMOUS &&
+            !cartState.loading
+          ) {
+            this.load(OCC_CART_ID_CURRENT, userId);
+          }
+          this.checkInitLoad = false;
+        }),
+        switchMapTo(cartSelector$),
+        // create cart can happen to anonymous user if it is empty or to any other user if it is loaded and empty
+        withLatestFrom(this.userIdService.getUserId()),
+        filter(([cartState, userId]) =>
+          Boolean(
+            userId === OCC_USER_ID_ANONYMOUS ||
+              cartState.success ||
+              cartState.error
+          )
+        ),
+        take(1)
+      )
+      .pipe(
+        tap(([cartState, userId]) => {
+          if (isEmpty(cartState.value)) {
+            this.multiCartFacade.createCart({
+              userId,
+              extraData: {
+                active: true,
+              },
+            });
+          }
+        }),
+        switchMapTo(cartSelector$),
+        filter((cartState) => Boolean(cartState?.success || cartState?.error)),
+        // wait for active cart id to point to code/guid to avoid some work on temp cart entity
+        withLatestFrom(this.activeCartId$),
+        filter(
+          ([cartState, cartId]) => !this.isCartCreating(cartState, cartId)
+        ),
+        map(([cartState]) => cartState.value),
+        filter((cart): cart is Cart => !isEmpty(cart)),
+        take(1)
+      );
```

Comment:
It's a bit hard to see the diff.  Can you help me understand the change? It looks like there are 2 new `pipe().` some new `?.` and `cart is Cart`.  ( all of which I presume must resolve an rxjs 7 issue)

Reply:
Correct, the pipe operator only supports up to 10 arguments before the typing gets too expensive to infer and falls back to `Observable<any>`. To keep the type information available, I broke the pipe arguments down into multiple chained calls with less than 10 arguments.  This is functionally equivalent to all pipe operators being in a single call.

The optional chaining operators are unnecessary and will be removed.

The observable type did not match the return type due to still including `undefined` in the type union.  I adjusted the `filter` function's return type to be a type guard to allow the `filter` to narrow the stream emit type.